### DISCUSSION TITLE
Compiler errors when specifying a target Java source file

### DISCUSSION THREAD

**Discussion by rafeememon** [bytedeco/javacpp#636] [Q&A] [answered]
I'm working on a JavaCPP wrapper around the [nanodbc](https://github.com/nanodbc/nanodbc) library for connecting to ODBC data sources from Java.  I'm bringing the library in via a git submodule and using gradle-javacpp to build the wrapper.  I'm using Windows 11 and Visual Studio 2022 Community.  I'm running into a couple issues, but I'm not exactly sure how/if they are related.

Repo: https://github.com/rafeememon/nanodbc-java-test/tree/f10465816f3fdae03553834d0d26353063a9e2db

I have a basic mapping configured in [NanodbcConfig.java](https://github.com/rafeememon/nanodbc-java-test/blob/f10465816f3fdae03553834d0d26353063a9e2db/src/main/java/io/nanodbc/NanodbcConfig.java).  In its current state, `./gradlew build` completes successfully, and `jninanodbc.dll` appears in the expected platform subdirectory under `build/classes`.  However, no wrapper .java file is generated to interface with the native library.

Adding a `target` in the `@Properties` annotation of the mapping by uncommenting [this line](https://github.com/rafeememon/nanodbc-java-test/blob/f10465816f3fdae03553834d0d26353063a9e2db/src/main/java/io/nanodbc/NanodbcConfig.java#L14) results in the wrapper .java file being generated, but I also encounter a slew of new compiler errors in the generated `jninanodbc.cpp` file.  It seems as though the errors are for illegal member accesses that should have been `#define`d away by my compiler options added to the build tasks [here](https://github.com/rafeememon/nanodbc-java-test/blob/f10465816f3fdae03553834d0d26353063a9e2db/build.gradle#L94-L99).

Build log without specifying `target`: [gradle_build_without_target.log](https://github.com/bytedeco/javacpp/files/10079710/gradle_build_without_target.log)
Build log with specifying `target`: [gradle_build_with_target.log](https://github.com/bytedeco/javacpp/files/10079713/gradle_build_with_target.log)

Is there something I should be doing differently to properly generated the wrapper .java file?  Thanks for any guidance you can provide!

**Comment by saudet** [comment, marked as answer]
We need to tell JavaCPP not to try to translate to variables any of those `#define` macros manually:
https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes#ignoring-attributes-and-macros
It's a bit tedious, but once it's done it doesn't usually take a lot of time to maintain.

**Reply by rafeememon** [reply]
@saudet thanks!  I'll give that a try.  I suppose I didn't realize that the generated `jninanodbc.cpp` file would be different if I specified a target.  Out of curiosity, how is the .dll intended to be used without the generated wrapper `target` class?

**Reply by saudet** [reply]
If that's a C API, you could use JNA, JNR, Panama, etc.

**Reply by rafeememon** [reply]
Thanks for the pointers.  I've made a lot of progress in getting my project to compile, but I'm running into another issue that's stumping me.  It seems that for the functions dealing with strings, either as parameters or returned values, the compiler isn't able to properly construct a `StringAdapter`.  I suspect I need to add a `.cast()` somewhere, but I'm not exactly sure.

Console output from `javacppBuildCompiler`: https://scans.gradle.com/s/ymjjizojn7ciu/console-log?task=:javacppBuildCompiler
Updated repo: https://github.com/rafeememon/nanodbc-java-test/tree/c9d591caf9cc545b659585f2310a2cd9ce70cf21

```
D:\git\nanodbc-java-test\build\classes\java\main\io\nanodbc\jniNanodbcNative.cpp(499): error C2446: ':': no conversion from 'const wchar_t [1]' to 'T *'
        with
        [
            T=char32_t
        ]
...
D:\git\nanodbc-java-test\build\classes\java\main\io\nanodbc\jniNanodbcNative.cpp(499): error C2665: 'std::basic_string<char32_t,std::char_traits<char32_t>,std::allocator<char32_t>>::basic_string': no overloaded function could convert all the argument types
```

The errors seem to come from the following constructor:

```
StringAdapter(const   signed   int* ptr, typename std::basic_string<T>::size_type size, void* owner) : ptr((T*)ptr), size(size), owner(owner),
        str2(ptr ? (T*)ptr : L"", ptr ? (size > 0 ? size : wcslen((wchar_t*)ptr)) : 0), str(str2) { }
```

**Reply by saudet** [reply]
Right, that's a tricky one. We need to get a pointer out of that.

**Reply by saudet** [reply]
I'm not seeing any "wchar_t" inside the header file. Where is it coming from?

**Reply by rafeememon** [reply]
That's a good hint... It looks like it was coming about because `nanodbc::string` is typedef'ed to `std::wstring`.  The mapping I needed to make the compiler happy was:
```
.put(new Info("nanodbc::string").cppTypes("std::wstring").annotations("@StdWString"))
```
Thanks for pointing me in the right direction!
